paddle/fluid/operators/math/softmax.cu

```diff
@@ -108,6 +113,120 @@ template class SoftmaxGradFunctor<platform::CUDADeviceContext, double>;
 template class SoftmaxGradFunctor<platform::CUDADeviceContext,
                                   platform::float16>;
 
+// NOTE: upgrade value type for float16 and float32
+static __device__ __forceinline__ float real_exp(platform::float16 x) {
```

Comment:
Should be called "cast_to_float32_exp". Not sure what does "real" mean

Reply:
this should be put into float16.h and extend the std namespace.

Reply:
@dzhwinter fp32 still need a way to return double result here.